arangod/Scheduler/ThreadPoolScheduler.h

```diff
@@ -0,0 +1,225 @@
+  // _queueRetryTime_us is the number of microseconds this particular
+  // thread should spin before going to sleep. Note that this spinning is only
+  // done if the thread has actually started to work on a request less than
+  // 1 seconds ago.
+  // _sleepTimeout_ms is the amount of ms the thread should sleep before waking
+  // up again. Note that each worker wakes up constantly, even if there is no
+  // work.
+  //
+  // All those values are maintained by the supervisor thread.
+  // Currently they are set once and for all the same, however a future
+  // implementation my alter those values for each thread individually.
+  //
+  // _lastJobStarted is the timepoint when the last job in this thread was
+  // started. _working indicates if the thread is currently processing a job.
+  //    Hence if you want to know, if the thread has a long running job, test
+  //    for _working && (now - _lastJobStarted) > eps
+  struct WorkerState {
+    uint64_t _queueRetryTime_us;  // t1
+    uint64_t _sleepTimeout_ms;    // t2
+    std::atomic<bool> _stop, _working, _sleeping;
+    // _ready = false means the Worker is not properly initialized
+    // _ready = true means it is initialized and can be used to dispatch tasks
+    // to _ready is protected by the Scheduler's condition variable & mutex
+    bool _ready;
+    std::atomic<clock::time_point> _lastJobStarted;
+    std::unique_ptr<ThreadPoolSchedulerWorkerThread> _thread;
+    std::mutex _mutex;
+    std::condition_variable _conditionWork;
+
+    // initialize with harmless defaults: spin once, sleep forever
+    explicit WorkerState(ThreadPoolScheduler& scheduler);
+    WorkerState(WorkerState const&) = delete;
+    WorkerState& operator=(WorkerState const&) = delete;
+
+    // cppcheck-suppress missingOverride
+    bool start();
+  };
+
+  std::unique_ptr<WorkItemBase> getWork(std::shared_ptr<WorkerState>& state);
+  void startOneThread();
+
+  // Check if we are allowed to pull from a queue with the given index
+  // This is used to give priority to "FAST" and "MED" lanes accordingly.
+  bool canPullFromQueue(uint64_t queueIdx) const noexcept;
+
+  void runWorker();
+
+  [[nodiscard]] bool queueItem(RequestLane lane,
+                               std::unique_ptr<WorkItemBase> item,
+                               bool bounded) override;
+
+ private:
+  NetworkFeature& _nf;
+  SharedPRNGFeature& _sharedPRNG;
+
+  std::atomic<uint64_t> _numWorkers;
+  std::atomic<bool> _stopping;
+  std::atomic<bool> _acceptingNewJobs;
+
+  // Since the lockfree queue can only handle PODs, one has to wrap lambdas
+  // in a container class and store pointers. -- Maybe there is a better way?
+  struct {
+    /// @brief the number of items that have been enqueued via tryBoundedQueue
+    /// Items that are added via an unbounded queue operation are not counted!
+    std::atomic<uint64_t> numCountedItems{0};
+    boost::lockfree::queue<WorkItemBase*> queue;
```

Comment:
boost lockfree queue performance is bad compared to other lockfree queue or even intrusive single linked list with mutex. also it contains data races
 Could you change this too?

Reply:
no. out of scope.